Element copying feature

mcapeta opened the pull request:
This PR enables selecting and copying existing content element (resolves #45) to the current activity.
It alse fixes #628 - element order issue within element selection dialog.

mcapeta committed:
Introduce element copying

mcapeta committed:
Fix allowed types definition for adding an element

mcapeta committed:
Scope `layout`flag to layout controls only

mcapeta committed:
Fix circular dependency issue

underscope committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Enable custom submit button label on `<select-element />`

mcapeta committed:
Merge branch 'feature/element-copying' of github.com:ExtensionEngine/tailor into feature/element-copying

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Improve/fix element copying

- avoid type/subtype overriding
- keep `refs` for copied objects
- code clean up

mcapeta committed:
Clean up

mcapeta committed:
Enable copying embedded elements in batch

mcapeta committed:
Typo

mcapeta committed:
Enable copying multiple questions

mcapeta committed:
[WIP] Enable copying multiple assessments

mcapeta committed:
Enable copying multipe assessments

in `AssessmentGroup`

mcapeta committed:
Enable copying multiple elements in `ContentContainer`

mcapeta committed:
Rename events & methods

to indicate support for saving multiple elements at once

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Change `SelectElement` dialog's icon

mcapeta committed:
Hide bottom sheet menu when element selection dialog is visible

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Ability to toggle element selection

mcapeta committed:
Clear element selection on `Back` action

mcapeta committed:
Change dialog heading

mcapeta committed:
Make copy element scrollable

to keep actions available regardless content size/height

mcapeta committed:
Fix assessment type resolution

mcapeta committed:
Clean up

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Revert `saveElement` for `ContentContainer`s

and print deprecation message

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,9 +2,26 @@
     :value="true"
-    header-icon="mdi-transit-connection-variant"
+    :header-icon="headerIcon"
+    scrollable
```
Order (after width)

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,9 +2,26 @@
-    <template v-slot:header>{{ heading }}</template>
+    <template v-slot:header>
+      {{ heading }}
+      <v-btn
```
Add slot instead and expose toggleSelectAll and allElementsSelected (creates error for element link as well).

mcapeta replied:
Seems that creating a slot just moves the problem to the consumers and requires repeatable code in them (for similar behaviour among them, of course). Instead, I've managed to handle toggling ability inside the component by checking whether it allows selecting multiple items and the content is loaded or not. Let me know if you see any drawbacks in this approach 😉.

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,9 +2,26 @@
+        v-if="selectedActivity && !loadingContent"
+        @click="toggleSelectAll"
+        text dark
+        class="float-right">
```
Or spacer / justify between

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,9 +2,26 @@
+        text dark
+        class="float-right">
+        <template v-if="allElementsSelected">
+          <v-icon class="mr-2">mdi-checkbox-multiple-blank-outline</v-icon>
```
Instead of repeating the same block consider creating a computed returning icon and label based on state.

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -51,15 +69,24 @@ export default {
-    allowedTypes: { type: Array, default: () => [] }
+    allowedTypes: { type: Array, default: () => [] },
+    submitLabel: { type: String, default: 'Save' },
+    headerIcon: { type: String, default: null }
```
Default to mdi-plus or brick representing the element add?

underscope commented on client/components/common/tce-core/AddElement/index.vue:
```diff
@@ -17,32 +17,42 @@
+    <v-bottom-sheet
+      @input="isVisible = $event"
+      :value="isVisible && !showElementBrowser"
+      max-width="1240" inset>
```
Move inset to the next line (inline only booleans)

underscope commented on client/components/common/tce-core/AddElement/index.vue:
```diff
@@ -17,32 +17,42 @@
-            Element width
-            <span class="px-1">{{ elementWidth }}</span>%
-          </div>
+          <v-btn
```
Does it make sense to create a slot and control add element visibility from outside (Move this component into AddNewElement and have a new index.vue wrapper that does add new / add existing handling)? ` :value="isVisible && !showElementBrowser"` looks a bit off 🤔

mcapeta replied:
Craves for some sort of separation 😄 , but I am kind of running in circles trying to clean this up... Have some ideas but I would like to resolve that in a separate PR. For now, I just moved some template and styles to `AddNewElement` component.

underscope commented on client/components/common/tce-core/AddElement/index.vue:
```diff
@@ -137,33 +167,41 @@ export default {
-      this.$emit('add', element);
+    addElements(elements) {
+      const items = elements.map((it, index) => {
+        const position = this.position + index;
```
if inserting multiple items between elements it needs to be bellow max as well ;)

mcapeta replied:
Reworked `calculatePosition` util with `AddElement`'s component logic, there were certain problems with inserting multiple elements I wasn't aware of 👀 .

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Resolve method name collision 🐂 🚜

mcapeta committed:
Resolve element width on appropriate node

mcapeta committed:
Clean up

attribute order

mcapeta committed:
Clean up

move button properties definition from tamplate to component logic

mcapeta committed:
Define default heading icon for `SelectElement` component

mcapeta committed:
Fix selection toggling in `SelectElement`

- disable for non-multiple selection
- allow all types by default
- fix selection format/interface violation

mcapeta committed:
[WIP] Clean up add element

Disengage adding existing and new elements

mcapeta committed:
Fix `allowedTypes` calculation

Use remapped items to conform supported types

mcapeta committed:
Refactor `calculatePosition` util

- enable calculating multiple positions
- expose additional helper

mcapeta committed:
Leverage `calculatePosition` improvement

Abandon `isFirstChild` flag

mcapeta committed:
Resolve developer idiocy 🐂

mcapeta committed:
Clean up

mcapeta committed:
Change event names from `camelCase` to `action:namespace` format

mcapeta committed:
Clean up

Rename method name to plural form

mcapeta committed:
Fix element preview layout

mcapeta committed:
Avoid element movement upon de/selection

mcapeta committed:
Move "Select All" button from modal header to body

kjuej commented:
@mcapeta @underscope 
Minor note: I'd like the copy more button to be more noticeable. This way it seems like more of a headline.
<img width="780" alt="Screen Shot 2020-11-12 at 11 58 06" src="https://user-images.githubusercontent.com/34127147/98931593-82c1a300-24de-11eb-8d81-5733f01350c9.png">

kjuej commented:
@mcapeta seems like i can only copy elements from 1 repo, is this by design?

kjuej commented:
@mcapeta don't show deselect/select button if there are no elements to copy.
<img width="703" alt="Screen Shot 2020-11-12 at 15 05 21" src="https://user-images.githubusercontent.com/34127147/98949865-911cb880-24f8-11eb-8f55-cc9d6efd4262.png">

kjuej commented:
@mcapeta unable to add an element in a custom container:
<img width="1547" alt="Screen Shot 2020-11-12 at 15 10 47" src="https://user-images.githubusercontent.com/34127147/98950426-45b6da00-24f9-11eb-80ff-39c2524c11a3.png">

underscope commented:
@kjuej Per conversation with Marin, he will proceed by adding the ability to select repo

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Swap width controls with "copy existing" button

mcapeta committed:
Hide "Select all" buton if there's nothing to select

mcapeta committed:
Change "Select all" button color

mcapeta committed:
Show element border in content preview

mcapeta committed:
Ensure `allowedTypes` is an array

mcapeta committed:
Map deprecated `saveElement` event to proper handler

mcapeta committed:
Fix/improve checks against `allowedTypes` list

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
[WIP] Create `<select-repository />` component

within `SelectElement` component

mcapeta committed:
Ability to select cross-repository elements

mcapeta committed:
Map deprecated event to proper one 🐂 🐛

mcapeta committed:
Clean up

Simplify element events emitting

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Scope relationships selection to the same repo

mcapeta committed:
Detach `<select-repository />` from store 🔧

mcapeta committed:
Clean up

mcapeta committed:
Merge branch 'develop' into feature/element-copying

bubafinder commented:
@mcapeta 
Disable editing elements while previewing/copying 

![image](https://user-images.githubusercontent.com/41724080/102087248-8e73f280-3e19-11eb-9b77-69db92a2a22d.png)

kjuej commented:
@mcapeta If a repo is empty (has no children), you're unable to enter it when in "copy element" menu:
![CopyElement](https://user-images.githubusercontent.com/34127147/102088916-d5fb7e00-3e1b-11eb-9424-a78d18bd2f02.gif)
Maybe a message should be added stating that the repo is empty once you enter it?

mcapeta committed:
Simplify repository select component

- use `<v-combobox />` instead of infinite scroll

mcapeta committed:
Improve empty set messages

mcapeta committed:
Revert empty result set check 🐛

mcapeta committed:
Disable filtering empty set

mcapeta committed:
Improve/clean up element selection

mcapeta committed:
Disable editting elements while selecting them

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Fix position calculation 🔧

mcapeta committed:
Default calculated position to the end of the array

mcapeta committed:
Fix `ADD_INTO` action position calculation

mcapeta committed:
Clean up

improve position resolution in embedded container

bubafinder commented:
Order of elements inside activity is not preserved 

![image](https://user-images.githubusercontent.com/41724080/102875648-5afe1d00-4444-11eb-83e0-0969a0b5546f.png)

bubafinder commented:
If I try to play the video, the element is selected. If some element has interaction (modal, video, audio, ...) don't select them on interaction click.

![image](https://user-images.githubusercontent.com/41724080/102876031-f0011600-4444-11eb-8915-2e8a85c46c54.png)

bubafinder commented:
Use default color for selected elements. Also, I don't prefer different right border color in relation to top, left and bottom color.

In this PR
![image](https://user-images.githubusercontent.com/41724080/102876423-74ec2f80-4445-11eb-9ff6-b4f124be729f.png)

Color that we use in activity
![image](https://user-images.githubusercontent.com/41724080/102876462-833a4b80-4445-11eb-957f-2cc4c4e21b3e.png)

bubafinder commented:
Expected behavior: If the select repository input field is cleared, I expect to see a list of all repositories.
Current behavior: If the select repository input field is cleared, I can see only the current repository in the list, and need to input a few chars to filter repositories.

![image](https://user-images.githubusercontent.com/41724080/102876577-af55cc80-4445-11eb-94b8-c5577c956f83.png)

bubafinder commented:
Don't like this view. Maybe it is better to disable entering into the empty activity.
Also, it would be good to use min-height for the Copy elements modal.

![image](https://user-images.githubusercontent.com/41724080/102877822-a5cd6400-4447-11eb-95b3-dbdb8642eb2f.png)

bubafinder commented:
Cant remove selected style of interacted elements with deselecting & mouse focus out.

![image](https://user-images.githubusercontent.com/41724080/102887219-d10b7f80-4456-11eb-8f11-b6ec5ac8f1f8.png)

markolovric commented:
maybe to add a color to "All selected" state?

![image](https://user-images.githubusercontent.com/41724080/102888232-a4586780-4458-11eb-926e-ff005d429e01.png)

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Add loader while fetching content

mcapeta committed:
Clear state properly upon activity deselection

mcapeta committed:
Override selected element border

mcapeta committed:
Increase "empty activity" message height

mcapeta committed:
Tweak elements' preview for `disabled` state

mcapeta committed:
Add link to element edit mode

within element selection (preview mode)

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Merge branch 'develop' into feature/element-copying

mcapeta committed:
Fix element order within `<select-element>` preview

mcapeta commented:
> Expected behavior: If the select repository input field is cleared, I expect to see a list of all repositories.
> Current behavior: If the select repository input field is cleared, I can see only the current repository in the list, and need to input a few chars to filter repositories.

It is desirable to avoid fetching all available repositories at once for performance reasons.

>Don't like this view. Maybe it is better to disable entering into the empty activity.
Also, it would be good to use min-height for the Copy elements modal.

Disabling user to enter an empty activity could also be performance issue since we would need to fetch entire repository content to filter out empty activities.

Although your suggestions lead to better experience, I think we could rely on the fact that content authors should know where they copy the content from and not implementing your proposals would not affect them too much (open for discussion 😄 ).

bubafinder commented:
Everything you mentioned makes sense and I think it is more important than my suggestions.

bubafinder commented:
I managed to run the delete action in the Video Playlist element

![image](https://user-images.githubusercontent.com/41724080/105832482-4d4e3100-5fc8-11eb-9942-4ae355c0ee71.png)

bubafinder commented:
Can't see the border-bottom on the last element. Add a small margin-bottom to the last element.

![image](https://user-images.githubusercontent.com/41724080/105832922-cc436980-5fc8-11eb-8d45-5412cf557ce6.png)

bubafinder commented:
TODO: 
- test add into/bellow/above
- test drag&drop

mcapeta committed:
Remove `native` modifier for binded listener 🔧

mcapeta commented:
> Can't see the border-bottom on the last element. Add a small margin-bottom to the last element.

Cannot reproduce 🤷‍♂️ 
Which browser do you use (this scrollbar looks strange)?

> I managed to run the delete action in the Video Playlist element

This has to be handled within element itself - video playlist is not part of the Tailor codebase, it is custom element and will be improved together with other elements (AFAIK element preview/edit mode are going to be reworked soon).

bubafinder commented:
> Which browser do you use (this scrollbar looks strange)?

I am using Chrome. In system preferences, I have activated the option to have always visible scroll bar (this is a must for QA, to see the worst case)

> This has to be handled within element itself - video playlist is not part of the Tailor codebase, it is custom element and will be improved together with other elements (AFAIK element preview/edit mode are going to be reworked soon).

👍

mcapeta committed:
Add margin to the bottom of content preview

bubafinder commented:
Keep mouse click style only above the "VIEW ELEMENTS" button

![image](https://user-images.githubusercontent.com/41724080/106140036-832d1a00-616e-11eb-88c4-66bf6a6d97a9.png)

bubafinder commented:
The content inside the view is not responsive

![image](https://user-images.githubusercontent.com/41724080/106140183-b66fa900-616e-11eb-82f3-ba53292964a2.png)

mcapeta committed:
Rename class

mcapeta committed:
Make element containers non-floating 🔧

abasic commented on client/components/common/ContentPreview/Element.vue:
```diff
@@ -1,19 +1,40 @@
-  <div @click="toggleSelection" class="element-preview-container">
+  <div
+    @click="toggleSelection"
+    :class="['element-preview-container', elementWidth]">
```
In Tailor the classes that are not dynamic are placed in the plain `class` attribute. Please change it here and everywhere else it was used.

abasic commented on client/components/common/ContentPreview/Element.vue:
```diff
@@ -40,17 +62,32 @@ export default {
   display: flex;
   position: relative;
+  margin: 1rem 0;
+  float: none;
```
Why is this `float` definition necessary?

mcapeta replied:
Along with this one, there is also an `elementWidth` class binding which sets the `float: left;` to this element. I've put it here to annul that property (since it results in 0 height parent).

abasic commented on client/components/common/SelectElement/SelectRepository.vue:
```diff
@@ -0,0 +1,48 @@
+        this.$emit('selected', repository);
+      }
+    },
+    fetchRepositories: throttle(loader(function (search) {
```
Is there a reason for using `throttle` instead of `debounce`?

mcapeta replied:
Thought we want function invocation on the leading edge 🤷‍♂️ ... Updating to be debounced...

abasic commented on client/components/common/SelectElement/SelectRepository.vue:
```diff
@@ -0,0 +1,48 @@
+    fetchRepositories: throttle(loader(function (search) {
+      if (!search || search.length < SEARCH_MIN_LENGTH) return;
+      return repositoryApi.getRepositories({ search }).then(repos => {
+        this.repositories = uniqBy([...this.repositories, ...repos], 'id');
```
If this is searching based on the specific search parameter, why is it not using just the new ones that have come in based on the search instead of keeping the old ones as well?

mcapeta replied:
Thought it may be nice to keep them all since they're already fetched. Will update to reject them, though 😉 .

abasic commented on client/components/common/SelectElement/SelectRepository.vue:
```diff
@@ -0,0 +1,48 @@
+      }
+    },
+    fetchRepositories: throttle(loader(function (search) {
+      if (!search || search.length < SEARCH_MIN_LENGTH) return;
```
Should the repositories list be cleared if no search is given or the search can not be run?

mcapeta replied:
Same as above, it seemed nice to keep existing ones. Updating...

abasic commented on client/components/common/SelectElement/SelectRepository.vue:
```diff
@@ -0,0 +1,48 @@
+  }),
+  methods: {
+    select(repository) {
+      if (this.repositories.includes(repository)) {
```
Just a small note on checking objects in arrays in this manner. I assume this works but it would not in the case the `v-combobox` did not return the object that was passed but another one it created. It is usually better to check if objects exist by relying on unique data it contains.

mcapeta replied:
👍  Thanks for noting that, I was relying on the fact that `v-combobox` works with references when `return-object` prop is set to `true`. I'll update it to check element existence using `id` as you suggested just to make sure we're not affected by any possible underlying changes 💪 .

abasic commented on client/components/common/SelectElement/index.vue:
```diff
@@ -93,11 +185,23 @@ export default {
+        selectedActivity: { id: activityId },
+        repository: { id: repositoryId }
+      } = this;
+      const params = { activityId, repositoryId };
```
This might just be me but I think this is not really readable. Lately, I have been personally avoiding such deep destructing just because it can be hard to read and track. Even more when there are renaming's. 
I think the below is more readable and even less lines of code:
```js
const params = {
  activityId: this.selectedActivity.id,
  repositoryId: this.repository.id
};
```

abasic commented on client/components/common/tce-core/AddElement/AddNewElement.vue:
```diff
@@ -0,0 +1,137 @@
+            :key="element.position"
+            @click.stop="$emit('add', [element])"
+            :disabled="!isAllowed(element.type)"
+            class="element">
```
It is not apparent from the class name that this is a button that can be interacted with. I recommend changing to `add-element` or something along those lines.

abasic commented on client/components/common/tce-core/AddElement/AddNewElement.vue:
```diff
@@ -0,0 +1,137 @@
+            :disabled="!isAllowed(element.type)"
+            class="element">
+            <v-icon v-if="element.ui.icon">{{ element.ui.icon }}</v-icon>
+            <h5 class="body-2">{{ element.name }}</h5>
```
An `h5` tag should not be inside of a button. It is not semantically correct. The thing to use in this case is `span` (and in most other cases 😄).

mcapeta replied:
Nice catch! I missed that whole section obviously - copy-pasting is dev's enemy 🤦‍♂️

abasic commented on client/components/common/tce-core/AddElement/AddNewElement.vue:
```diff
@@ -0,0 +1,137 @@
+          <span>{{ group.name }}</span>
+        </div>
+        <div class="group-elements">
+          <button
```
Is there a reason not to use the `v-btn` component here?

mcapeta replied:
^^ Copy pasted, to my shame 😞 .

abasic commented on client/components/common/tce-core/AddElement/index.vue:
```diff
@@ -17,63 +17,70 @@
+          <v-btn
+            @click="showElementBrowser = !showElementBrowser"
+            dark text
+            class="">
```
Please remove this empty attribute.

mcapeta replied:
Sorry for wasting your time on this one 😱 .

abasic commented on client/components/common/tce-core/AddElement/index.vue:
```diff
@@ -17,63 +17,70 @@
+import reject from 'lodash/reject';
 import uuid from '@/utils/uuid';
 
+const SelectElement = () => import('components/common/SelectElement');
```
@ExtensionEngine/tailor Should these be defined here at the beginning, along with the other ones or directly in the `components` prop of the component? I am raising this since it is the first time I see this in Tailor and am wondering what is considered the better approach?

mcapeta replied:
~Linter says it's right in its place 😄 .~

E: Misread your comment. I'm voting for keeping it close to the regular imports and keeping `components` property concise.

abasic commented on client/components/common/tce-core/AddElement/index.vue:
```diff
@@ -137,33 +152,42 @@ export default {
+      if (!layout) return include || [];
+      const allowedElements = elementWidth === DEFAULT_ELEMENT_WIDTH
+        ? elements
+        : reject(elements, 'ui.forceFullWidth');
```
Why is this necessary?

mcapeta replied:
If user is willing to add a half width element it's supposed to be allowed to copy only those elements which support half width preview (and are not necessarily half-width elements originally).

abasic commented on client/components/common/tce-core/AddElement/index.vue:
```diff
@@ -174,106 +198,14 @@ export default {
+    },
+    show(val) {
+      if (val) return this.showElementPicker();
+      this.onHidden();
```
```suggestion
      return val ? this.showElementPicker() : this.onHidden();
```

abasic commented on client/components/common/tce-core/AddElement/index.vue:
```diff
@@ -137,33 +152,42 @@ export default {
-      return get(element, 'ui.forceFullWidth', false);
+    buildElement(el) {
+      const { processedWidth: width, activity } = this;
+      const { position, subtype, data = {}, initState = () => ({}) } = el;
```
@ExtensionEngine/tailor I think we should make sure that `initState` and `data` have a default value so that it is is not necessary to default here.

abasic commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,88 +2,180 @@
+        : await this.fetchActivities(repository);
+    },
+    fetchActivities: loader(function (repository) {
+      return repositoryApi.get(repository.id, { withActivities: true })
```
There exists an[ activity list route](https://github.com/ExtensionEngine/tailor/blob/develop/server/activity/index.js#L15) in Tailor that already fetches them based on the repository. Is there a reason why that was not used?

mcapeta replied:
No reason, I wasn't aware of it. Thanks!

abasic commented on client/utils/calculatePosition.js:
```diff
@@ -1,47 +1,48 @@
+};
+
+function getDeprecationWarning(config) {
+  if (!Object.prototype.hasOwnProperty.call(config, 'isFirstChild')) return;
```
Is there a reason why this is not written by just calling `hasOwnProperty` on the `config` object like so:
```js
if (!config.hasOwnProperty('isFirstChild')) return;
```

mcapeta replied:
It's not me, [linter rule](https://eslint.org/docs/rules/no-prototype-builtins) made me to 😄  .

abasic commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,88 +2,180 @@
-      }));
+      this.contentContainers = containers.map(container => {
+        const containerElements = elements.filter(it => it.activityId === container.id);
+        return { ...containers, elements: sortBy(containerElements, 'position') };
```
I think it should be `container` instead of `containers`.

mcapeta replied:
![bullseye](https://media.giphy.com/media/Z61gwKI72Gx2g/giphy.gif)

abasic commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,88 +2,180 @@
+        return { ...containers, elements: sortBy(containerElements, 'position') };
+      });
+    },
+    assignActivity(element) {
```
This is used when the element is being toggled on or off which can theoretically happen frequently. Is there a reason this can not be called on all of the elements right after they are fetched and ordered by position?

mcapeta replied:
Thank you 😉

mcapeta committed:
Clean up

separate static and dynamic class bindings

mcapeta committed:
Clean up & bugfix

- use `container` instead of `containers` as mapping result
- reduce number of activity assignments

mcapeta committed:
Repository selection component improvements

mcapeta committed:
Clean up

variable assignment readability

mcapeta committed:
Clean up

class name semantics

mcapeta committed:
Clean up

- use Vuetify `v-button` instead of HTML `button` directly
- do not use `h5` inside a `button`

mcapeta committed:
Remove empty `class` attr

mcapeta committed:
Clean up

use ternary over the early return when possible

mcapeta committed:
Leverage `activity` endpoint instead of modifing `repository`

mcapeta committed:
Polish treeview hover state cursor

mcapeta committed:
Change element selection

to toggle on checkbox click only

underscope commented on client/components/common/ContentPreview/Element.vue:
```diff
@@ -1,19 +1,40 @@
+              @click.stop="$emit('element:open', element.uid)"
+              color="pink darken-1"
+              fab small dark
+              :class="{ visible: hover }"
```
Bindings order

underscope commented on client/components/common/ContentPreview/Element.vue:
```diff
@@ -40,17 +62,32 @@ export default {
   display: flex;
   position: relative;
+  margin: 1rem 0;
+  float: none;
```
Order (positioning before margins)

underscope commented on client/components/common/SelectElement/SelectActivity.vue:
```diff
@@ -32,33 +34,37 @@
   computed: {
-    ...mapGetters('repository', ['activities']),
     groupedSelection: vm => groupBy(vm.selectedElements, 'outlineId'),
+    expandedItems: vm => map(vm.activities, 'id'),
```
Consider adding Ids in the name (i.e. `expandedActivityIds`)

underscope commented on client/components/common/tce-core/EmbeddedContainer.vue:
```diff
@@ -1,6 +1,6 @@
 <template>
   <element-list
-    @add="addItem"
+    @add="addItems"
```
Backwards comp 🤔 ?

mcapeta replied:
Updated to support both single item or an array of items. Should we deprecate single value signature or support both.

Nice catch btw, I just missed the `list-add` slot within `<element-list />`. Writing down all the things that can be customized within Tailor 😄

underscope replied:
Lets support both 👍

underscope commented on client/components/common/tce-core/EmbeddedContainer.vue:
```diff
@@ -21,12 +21,12 @@
 </template>
 
 <script>
+import calculatePosition from 'utils/calculatePosition';
```
🚨 calculatePosition should be part of tce-core (only tce-core and third party packages should be used within tce core; leftover; will be extracted into separate package). Not a must since it was like that before.

underscope commented on client/utils/calculatePosition.js:
```diff
@@ -1,47 +1,48 @@
  *     should be placed in relation to `newPosition`. Defaults to `REORDER` in
  *     case of items reorder.
- * @return {number} Calculated item position.
+ * @param {number} count A number of positions to return.
```
A number of items with changed position / for which position needs to be determinated ?

underscope commented on client/utils/calculatePosition.js:
```diff
@@ -1,47 +1,48 @@
+  const arr = [...items];
+  if (action === REORDER) arr.splice(newPosition, count);
+  let index = items.length;
+  if (newPosition !== undefined) {
```
Consider documenting logic / scenarios for each case

underscope commented on client/utils/calculatePosition.js:
```diff
@@ -1,47 +1,48 @@
- * @param {number} newPosition The position of the anchor item by which the new
- *     position will be calculated.
+ * Calculates item position(s) based on the options provided.
+ * @param {number} newPosition The index of the anchor item by which the new
```
For add after sake position might be a better name with new changes (can do a improvement PR ?; so it is easier to revisit)

underscope commented on client/utils/calculatePosition.js:
```diff
@@ -1,47 +1,48 @@
- * @return {number} Calculated item position.
+ * @param {number} count A number of positions to return.
+ * @return {(number|Array)} Single position if `count` is 1 or an array containing
+ * `count` positions.
```
Indent ?

underscope commented on client/store/modules/repository/activities/actions.js:
```diff
@@ -52,11 +52,9 @@ const clone = ({ commit }, mapping) => {
 
 const calculateInsertPosition = ({ state }, { activity, anchor, action }) => {
   const items = getOutlineChildren(state.items, activity.parentId);
+  if (action === ADD_INTO) return calculatePosition({ items, action });
```
Rather than calling calculatePosition twice determine payload conditionally (it can initially be set to `{ items, action }`) and if `!ADD_INTO` set `newPosition`

underscope commented on client/components/editor/Sidebar/ElementSidebar/ElementMeta/Relationships/Item.vue:
```diff
@@ -32,12 +32,14 @@
-      :allowed-types="allowedTypes" />
+      :allowed-types="allowedTypes"
+      use-current-repo
+      header-icon="mdi-transit-connection-variant" />
```
Order

underscope commented on client/components/editor/Sidebar/ElementSidebar/ElementMeta/Relationships/Item.vue:
```diff
@@ -84,6 +86,14 @@ export default {
+    select(elements) {
+      const items = elements.map(it => {
+        if (!it.activity) return it;
+        const { id, activity, activityId: containerId } = it;
```
We should probably rename activity or activityId on element (because now it looks like `activityId === activity.id`). Thoughts? 🤔

mcapeta replied:
It is confusing, though. If we're going to change what's preserved in db we should migrate data, aren't we? If you thought of renaming properties on FE only, I would personally leave it as is in one place or move it [here](https://github.com/ExtensionEngine/tailor/pull/624/files#diff-62af149e0cab2f7e0067b49fd93a40766741bf76f511f7f476577a3e6abaf1b5R141) where this data actually come from (but still keep in one place).

And if you have any better idea please let me know, maybe I just misunderstood your suggestion 🤷‍♂️ .

underscope replied:
I was thinking that for readability sake of this component within processor that is hydrating state of this component (component that you linked) we do the rename that you did here: `activityId => containerId` (and/or `activity => outlineActivity`) so this is more readable and we know which activity id represents which activity (one is outline parent and other one is container activity). Is good for now, consider it as an improvement.

underscope commented on client/components/editor/ActivityContent/ContainerList.vue:
```diff
@@ -12,15 +12,15 @@
-      @addSubcontainer="save"
-      @updateSubcontainer="update"
-      @deleteSubcontainer="requestContainerDeletion"
+      v-on="deprecatedListeners"
```
🏆

mcapeta committed:
Improve respository selection

underscope commented on client/components/editor/ActivityContent/ContainerList.vue:
```diff
@@ -91,26 +118,25 @@ export default {
+    saveContentElements(elements) {
+      castArray(elements).forEach(element => {
+        const elementChannel = this.$radio.channel(`element:${getElementId(element)}`);
+        return this.saveElement(element).then(() => {
```
Not sure if we should trigger multiple save notifications in this case. Thoughts 🤔 ?

mcapeta replied:
Reworked to show one notification once all elements are successfully saved. Next step is to save element in batch, as discussed (another PR).

underscope commented on client/components/content-containers/tcc-exam/edit/AssessmentGroup.vue:
```diff
@@ -106,17 +106,19 @@ export default {
-      if (assessment.id) return this.$emit('updateElement', assessment);
-      this.$emit('saveElement', assessment);
+      if (assessment.id) return this.$emit('update:element', assessment);
+      this.$emit('save:element', assessment);
```
Resolve event name using ternary ?

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,88 +2,176 @@
+    headerIcon: { type: String, default: 'mdi-toy-brick-plus-outline' },
+    useCurrentRepo: { type: Boolean, default: false }
+  },
+  data: () => {
```
Could be written more compact / less indented  `() => ({.....`

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,88 +2,176 @@
+      contentContainers: [],
+      selectedElements: [],
+      loadingContent: false,
+      activities: []
```
Consider grouping selected and fetched props (i.e. repository, activities, contentContainers, selectedActivity, selectedElements)

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,88 +2,176 @@
+    }),
+    allElementsSelected: vm => vm.selectedElements.length === vm.elements.length,
+    showBackButton: vm => vm.useCurrentRepo ? !!vm.selectedActivity : !!vm.repository,
+    containers() {
```
having contentContainers and containers is a bit .... Consider renaming this to processedContainers or similar (i.e. other one to selectedActivityContainers) :bowtie:

mcapeta committed:
Revert element interaction within preview

mcapeta committed:
Clean up 🐑 🐑 🐑 🐑 🐂 🧙

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -93,11 +181,22 @@ export default {
     close() {
       this.$emit('close');
+    },
+    openElement(elementId) {
```
goTo / openInEditor rather then openElement?

underscope commented on client/components/common/tce-core/AddElement/AddNewElement.vue:
```diff
@@ -0,0 +1,123 @@
+  },
+  methods: {
+    isAllowed(type) {
+      if (!this.allowedTypes.length) return true;
```
Ternary  or || 🤔  ?

underscope commented on client/components/common/tce-core/AddElement/index.vue:
```diff
@@ -17,63 +17,67 @@
+                <v-icon>mdi-select-compare</v-icon>
+              </v-btn>
+            </v-btn-toggle>
+            <v-divider class="mr-3" vertical />
```
Order

underscope commented on client/components/common/tce-core/AddElement/index.vue:
```diff
@@ -82,23 +86,31 @@ const ELEMENT_GROUPS = [
 
+const getQuestionData = (element, type) => {
+  const data = { width: LAYOUT.FULL_WIDTH };
+  const question = [{ id: uuid(), data, type: 'JODIT_HTML', embedded: true }];
```
Need to create improvement issue for this one, it shouldn't be here... (Ignore for now)

underscope commented:
ty ❤️ 🏆

mcapeta committed:
Clean up

- renaming variables
- attr order
- ternary or logical operators over early return (where possible)

mcapeta committed:
Support both one and multiple items to be added

mcapeta committed:
Polish `calculatePosition` helper's docs

mcapeta committed:
Bugfix: `selectedElements` initial value

mcapeta committed:
Ensure one notification per action (multiple elements saving)

MiroDojkic commented on client/components/common/ContentPreview/Element.vue:
```diff
@@ -1,19 +1,40 @@
+          :element="element"
+          :set-width="false"
+          class="content-element"
+          :class="{ selected: isSelected }" />
```
Please switch prop order (bound props first).

MiroDojkic commented on client/components/common/ContentPreview/Element.vue:
```diff
@@ -1,19 +1,40 @@
+              color="pink darken-1"
+              fab small dark
+              class="open-element-button"
+              :class="{ visible: hover }">
```
Sorry for nitpicking but I think the order of bound, static, boolean and then class props would be consistent with the rest of the codebase. 🙃

MiroDojkic commented on client/components/common/ContentPreview/Element.vue:
```diff
@@ -1,19 +1,40 @@
 <template>
-  <div @click="toggleSelection" class="element-preview-container">
+  <div :class="elementWidth" class="element-preview-container">
```
Please consider if this approach would make it more straightforward to infer what's going on with the layout straight from reading the template:
```suggestion
  <div :class="col-xs-`${elementWidth} float-none`" class="element-preview-container">
```

MiroDojkic commented on client/components/common/ContentPreview/Element.vue:
```diff
@@ -59,4 +96,20 @@ export default {
+
+.element-wrapper {
+  position: relative;
+  flex-grow: 1;
```
Consider using vuetify's `flex-grow-1` util class for this.

MiroDojkic commented on client/components/common/SelectElement/SelectActivity.vue:
```diff
@@ -32,33 +34,37 @@
+      if (!activities.length) return 'Empty repository';
+      if (!search || !$refs) return '';
+      const { excludedItems, nodes } = $refs.treeview;
+      if (excludedItems.size !== Object.keys(nodes).length) return '';
```
Would assigning this condition to boolean variable make it a bit simpler to read?
I.e.
```suggestion
      const hasSearchResults = excludedItems.size !== Object.keys(nodes).length;
      if (excludedItems.size !== Object.keys(nodes).length) return '';
```

MiroDojkic commented on client/components/common/SelectElement/SelectActivity.vue:
```diff
@@ -81,5 +87,11 @@ export default {
     border-radius: 12px !important;
   }
+
+  ::v-deep .v-treeview-node--leaf {
+    > .v-treeview-node__root, .v-treeview-node__content > * {
+      cursor: auto;
+    }
+  }
```
Do you think this would make the selector easier to read or is it too cryptic? 🤔 
```suggestion
  ::v-deep .v-treeview-node {
    &--leaf > &__root, &--leaf > &__content > * {
      cursor: auto;
    }
  }
```

mcapeta replied:
I wouldn't say it's cryptic, I like it. It is clear enough we're updating cursor behaviour somewhere within `.v-treeview-node` and readability of other selectors is pretty much the same in both cases so I pick concise one 😉 .

MiroDojkic commented on client/components/common/SelectElement/SelectRepository.vue:
```diff
@@ -0,0 +1,48 @@
+        this.$emit('selected', repository);
+      }
+    },
+    fetchRepositories: debounce(loader(function (search) {
```
Is the name `fetch` good enough as in the case of `select` method - if the component's context makes it easy to figure out what it does?
If not, should we rename `select` to `selectRepository`?

MiroDojkic commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,88 +2,174 @@
-    allowedTypes: { type: Array, default: () => [] }
+    submitLabel: { type: String, default: 'Save' },
+    headerIcon: { type: String, default: 'mdi-toy-brick-plus-outline' },
+    useCurrentRepo: { type: Boolean, default: false }
```
Consider renaming to `onlyCurrentRepo` or something similar to indicate that it's a boolean type of value.

mcapeta replied:
Initial version was `useOnlyCurrentRepo`. Seems I rejected the wrong one 😄 . Updating as suggested...

MiroDojkic commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,88 +2,174 @@
+      const { selectedActivity, activities } = this;
+      if (!selectedActivity || !activities.length) return [];
+      const rootTypes = getContainerTypes(selectedActivity.type);
+      let containers = activities.filter(({ type, parentId }) => {
```
Please consider moving arrow functions to component methods to make this computed prop more straightforward to read.

MiroDojkic commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,88 +2,174 @@
+          .filter(it => it.activityId === container.id)
+          .map(element => ({ ...element, activity }));
+        return { ...container, elements: sortBy(containerElements, 'position') };
+      });
```
It might be worth to abstract adding elements to a container here:
```suggestion
      this.contentContainers = processedContainers.map(this.addContainerElements);
```

mcapeta replied:
Some more context variables needed (`activity`, `elements`) so please take a look if you find new solution more readable 😉 .

MiroDojkic commented on client/components/common/mixins/deprecation.js:
```diff
@@ -0,0 +1,21 @@
+    deprecateEvent(handler, { oldEvent, newEvent, adaptArgs }) {
+      newEvent = newEvent || noCase(oldEvent, { delimiter: ':' });
+      const context = this;
+      return function () {
```
Why not use arrow function and avoid the need to bind this?

mcapeta replied:
Because of `arguments` binding. Arrow functions does not have them (i.e. they access arguments of the first non-arrow parent) if I'm not wrong. Leveraging them here to be able to decorate methods with different signatures...
I'll check it one more time before I resolve this one...

MiroDojkic commented on client/components/common/tce-core/AddElement/AddNewElement.vue:
```diff
@@ -0,0 +1,122 @@
+            <v-btn
+              @click.stop="$emit('add', [element])"
+              :disabled="!isAllowed(element.type)"
+              :color="hover ? 'pink' : '#333'"
```
Should we use `secondary` instead of `pink`, in case we ever decide to change palette?

MiroDojkic commented on client/components/common/tce-core/AddElement/AddNewElement.vue:
```diff
@@ -0,0 +1,122 @@
+  padding-top: 1.25rem;
+  font-size: 1rem;
+  font-weight: 500;
+  line-height: 1.75rem;
```
Do we want to leave this to change relative to font size (in the future)? If that's the case, we could use `1.75;` or `1.75em;`.

MiroDojkic commented on client/components/editor/ActivityContent/ContainerList.vue:
```diff
@@ -91,26 +120,32 @@ export default {
+      Promise.map(contentElements, element => {
+        const elementChannel = this.$radio.channel(`element:${getElementId(element)}`);
+        return this.saveElement(element)
+          .then(() => elementChannel.emit('saved'))
+          .catch(err => {
+            elementChannel.emit('error', err);
+            return Promise.reject(err);
+          });
```
Can we flatten this promise chain by moving getting channel and emitting event through channel to `saveElement` method?

mcapeta replied:
> Here's a couple of small suggestions (mostly format-related), please see if any of these make any sense. 😄

Thank you 😉

mcapeta committed:
Namespace data properties

Separate loaded items from selected ones

MiroDojkic reviewed (commented):
Here's a couple of small suggestions (mostly format-related), please see if any of these make any sense. 😄

mcapeta committed:
Update `calculatePosition` docs

mcapeta committed:
Clean up

- binding order
- renaming stuff
- simplifying statements

mcapeta committed:
Clean up

mcapeta committed:
Clean up

mcapeta commented:
> Here's a couple of small suggestions (mostly format-related), please see if any of these make any sense. 😄

Thank you 😉

mcapeta committed:
Bugfix: Bind deprecated events to corresponding handlers correctly

mcapeta committed:
Clean up

mcapeta committed:
🐂

mcapeta committed:
Chasing own tail...

mcapeta committed:
Clean up container processing

Extract helpers to methods

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,102 +2,212 @@
-    selectedActivity: null,
-    contentContainers: [],
-    selectedElements: [],
+    items: {
```
Grouping this under items seems bit off/too generic. Consider improving this (separate PR); if selection is used to group the selected, should these be something like `selectionOptions` ? 🤔

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,102 +2,212 @@
+      ]);
+      return flatMap(containers, it => [it, ...this.getSubcontainers(it)]);
+    },
+    elements() {
```
processedElements 🤔  / elementSelectionOptions  🤡   ?

underscope replied:
Same as prev one, it can be just a function which will assign to items.elements / selectionOptions.elements. Then some other aspects will be more readable like:
`vm.selection.elements.length === vm.elements.length`
will become =>
`vm.selection.elements.length === vm.selectionOptions.elements.length`.
Can filtering of allowedTypes be moved to => assignElements? That one should probably be more descriptive as well 🤔

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,102 +2,212 @@
+      const type = this.selection.activity?.type;
+      return type && getContainerTypes(type);
+    },
+    processedContainers() {
```
It might be just a function (we have already a lot of `something` containers on the context. It is only used within a single function so it can be eliminated; Separate PR)

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,102 +2,212 @@
+      return elements.filter(it => this.allowedTypes.includes(it.type));
+    },
+    toggleButton() {
+      const { allElementsSelected, elements, multiple, selection } = this;
```
🔝 This one as well; see prev comment; elements => selectionOptions.elements

underscope commented on client/components/common/SelectElement/index.vue:
```diff
@@ -2,102 +2,212 @@
+    }
+  },
   methods: {
+    getTypePosition({ type }) {
```
Not sure if it makes sense here since it is only used once and is one liner.
